File: iOS/DoraemonKitDemo/DoraemonKitDemo/DemoVC/MemoryLeak/DoraemonDemoMemoryLeakView.m
//
//  DoraemonDemoMemoryLeakView.m
//  DoraemonKitDemo
//
//  Created by didi on 2019/10/6.
//  Copyright © 2019 yixiang. All rights reserved.
//

#import "DoraemonDemoMemoryLeakView.h"
#import "DoraemonDemoMemoryLeakModel.h"

@interface DoraemonDemoMemoryLeakView()

@property (nonatomic, strong) DoraemonDemoMemoryLeakModel *model;

@end

@implementation DoraemonDemoMemoryLeakView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor redColor];
    
        _model = [[DoraemonDemoMemoryLeakModel alloc] init];
        [_model addBlock:^{
            [self printXX];
        }];
        [_model install];
    }
    return self;
}

- (void)printXX {
    NSLog(@"view XX");
}

- (void)dealloc {
    NSLog(@"DoraemonDemoMemoryLeakView dealloc");
}

@end

File: iOS/Swift/DoKitSwiftDemo/DoKitSwiftDemo/MemoryLeak/DoraemonDemoMemoryLeakView.swift
//
//  DoraemonDemoMemoryLeakView.swift
//  DoKitSwiftDemo
//
//  Created by didi on 2020/5/19.
//  Copyright © 2020 didi. All rights reserved.
//

import UIKit

class DoraemonDemoMemoryLeakView: UIView {
    var model : DoraemonDemoMemoryLeakModel
    override init(frame: CGRect) {
        model = DoraemonDemoMemoryLeakModel()
        super.init(frame: frame)
        self.backgroundColor = UIColor.orange
        model.closure = { ()->Void in
            self.doSomeThing()
        }
        model.callClosure()
    }
    
    func doSomeThing() {
        print("view doSomeThing");
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        print("view deinit")
    }
    
}
